chapter_recurrent-neural-networks/language-models-and-dataset.ipynb

```diff
@@ -0,0 +1,699 @@
+    "\\end{aligned}\n",
+    "$$\n",
+    "\n",
+    "The probability formulae that involve one, two, and three variables are typically referred to as *unigram*, *bigram*, and *trigram* models, respectively. In the following, we will learn how to design better models.\n",
+    "\n",
+    "## Natural Language Statistics\n",
+    "\n",
+    "Let us see how this works on real data.\n",
+    "We construct a vocabulary based on the time machine dataset as introduced in :numref:`sec_text_preprocessing` \n",
+    "and print the top 10 most frequent words.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%mavenRepo snapshots https://oss.sonatype.org/content/repositories/snapshots/\n",
+    "\n",
+    "%maven ai.djl:api:0.10.0-SNAPSHOT\n",
+    "%maven org.slf4j:slf4j-api:1.7.26\n",
+    "%maven org.slf4j:slf4j-simple:1.7.26\n",
+    "\n",
+    "%maven ai.djl.mxnet:mxnet-engine:0.10.0-SNAPSHOT\n",
+    "%maven ai.djl.mxnet:mxnet-native-auto:1.7.0-backport"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load ../utils/plot-utils\n",
+    "%load ../utils/PlotUtils.java\n",
+    "%load ../utils/TimeMachineUtils.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load ../utils/TimeMachineUtils.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ai.djl.ndarray.*;\n",
+    "import ai.djl.ndarray.types.*;\n",
+    "import ai.djl.ndarray.index.*;\n",
+    "import ai.djl.util.Pair;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "NDManager manager = NDManager.newBaseManager();"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "String[][] tokens = TimeMachine.tokenize(TimeMachine.readTimeMachine(), \"word\");\n",
+    "// Since each text line is not necessarily a sentence or a paragraph, we\n",
+    "// concatenate all text lines\n",
+    "List<String> corpus = new ArrayList<>();\n",
+    "for (int i = 0; i < tokens.length; i++) {\n",
+    "    for (int j = 0; j < tokens[i].length; j++) {\n",
+    "        if (tokens[i][j] != \"\") {\n",
+    "            corpus.add(tokens[i][j]);\n",
+    "        }\n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "Vocab vocab = new Vocab(new String[][] {corpus.toArray(new String[0])}, -1, new String[0]);\n",
+    "for (int i = 0; i < 10; i++) {\n",
+    "    Map.Entry<String, Integer> token = vocab.tokenFreqs.get(i);\n",
+    "    System.out.println(token.getKey() + \":\" + token.getValue());   \n",
```

Comment:
Would this look better with ": " instead of ":" (add a space)?

Reply:
Yep. Also changed all the other outputs in the section to look like this. I was using built in toString() of `Map.Entry` for the others so now they all look the same.